woocommerce/woocommerce
# Add endpoint to get SSR

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:


This PR introduces an API endpoint that allows WooCommerce.com to obtain SSR remotely. This will be helpful for support interactions, allowing to skip manual work copy/pasting SSR report into ticket forms.

Link to thread: p6JqRr-8v3-p2



Closes # .



## Preconditions
- Connect the store to your account on WooCommerce.com from **WP admin -> WooCommerce -> Extensions -> My subscriptions.**. Ensure you perform the connection from behalf of user with admin or shop managing role.


## Test an unauthenticated request
1. Send a request to the endpoint (ensure to replace domain with your site domain):
```bash
curl "https://wp.test:4443/wp-json/wccom-site/v1/ssr?token=AAA&signature=BBB"
```
2. Confirm you have received a 401 Unauthorized response with body
```json
{
	"code": "invalid_token",
	"message": "Invalid access token provided",
	"data": {
		"status": 401
	}
}
```

## Test SSR submission failure
1. Comment out [lines 102-110](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L102-L110) and [114-122](https://github.com/woocommerce/woocommerce/blob/2800be9b7bae98b3d43247948671e63a96edb234/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php#L114-L122) in file `/plugins/woocommerce/includes/wccom-site/class-wc-wccom-site.php` to bypass validation of authentication credentials.
2. Retry the same request from the previous section.
3. Confirm you received a 200 response signifying that the request has been processed, but submission of SSR data has failed with 404 error (as WooCommerce.com's endpoint is not publicly available yet).
```json
{
	"success": false,
	"error_code": "failed_submitting_ssr",
	"error_message": "Submitting SSR data failed with response code: 404"
}
```

## Test successful SSR submission
1. Go to https://webhook.site/, which is an online tool providing with a publicly accessible endpoint for testing. Copy the URL it generated.  <img width="809" alt="image" src="https://github.com/woocommerce/woocommerce/assets/2803994/3259f0bd-f335-4bdd-a142-09f34db45366">
2. Click Edit button in the top right corner and change the default status code to 201 
![image](https://github.com/woocommerce/woocommerce/assets/2803994/8da8d360-d6fa-4762-91aa-96d1aa3f23dd)
3. Add the below code to your site, e.g., to `mu-plugins`. It will point Woo Helper requests to the endpoint provided by webhook.site (replace `https://webhook.site/11179c2e-b82e-49ae-bc71-jk23kd23lk2n` with your own link):
```php
add_filter( 'woocommerce_helper_api_base', function( $base ) {
	return 'https://webhook.site/11179c2e-b82e-49ae-bc71-jk23kd23lk2n';
} );
```
4. Retry the same request run in the previous sections:
```bash
curl "https://wp.test:4443/wp-json/wccom-site/v1/ssr?token=AAA&signature=BBB"
```
5. Confirm you receive a 200 response stating SSR submission succeeded.
```json
{
	"success": true,
	"message": "SSR data submitted successfully"
}
```
6. Go back to webhook.site page in browser, and confirm it shows that it received a POST request. Scroll to the bottom of page and confirm it shows SSR data in JSON format: 
<img width="626" alt="image" src="https://github.com/woocommerce/woocommerce/assets/2803994/ad3285d4-6699-4832-bd1c-b473fdc93d63">
